<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Quinn W" />


<title>Basic VCF Analysis</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Analysis of Genetic Diversity in Ciliates</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="Heterozygosity_Tutorial.html">Heterozygosity Tutorial</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Basic VCF Analysis</h1>
<h4 class="author">Quinn W</h4>
<h4 class="date">5/21/2020</h4>

</div>


<p>First, we’ll need to load the needed libraries and read the vcf data in as a vcfR object. For the plots, we will want to work with the vcf as a data frame, so we can use the function vcfR2tidy to extract the tidy data frame from the vcfR object.</p>
<pre class="r"><code>#file path to vcf

spirostomum_file_668_to_668 &lt;- &quot;./Spirostomum/mappedto668/LKH668_FPE_q24_minlen125.fastq.gz_LKH668_Filtered.UC.Final.NTD.ORF.fasta_deduped_filtered.vcf.recode.vcf&quot;

#create vcfR object

spirostomum_vcfR_668_to_668 &lt;- read.vcfR(spirostomum_file_668_to_668)

#extract tidy data frame

spirostomum_668_to_668 &lt;- vcfR2tidy(spirostomum_vcfR_668_to_668, 
                             single_frame = TRUE, 
                             info_types = TRUE, 
                             format_types = TRUE)[[&quot;dat&quot;]]</code></pre>
<p>The following functions can be used to better understand the vcf file you’re analyzing.</p>
<pre class="r"><code>#find number of polymorphisms per contig, then graph the distribution of the number of polymorphisms per contig
#takes tidy vcf data and optional arguments for title and fill

graph_poly_per_contig &lt;- function(data, 
                                  my_title = &quot;Polymorphisms per Contig&quot;, 
                                  fill = &quot;#505D83&quot;) {
  data %&gt;% 
    group_by(CHROM) %&gt;%
    summarize(num_poly = n()) %&gt;%
    ggplot(aes(x = num_poly)) + 
    geom_histogram(binwidth = 1, fill = fill) +
    theme_minimal() +
    labs(title = my_title,
         x = &quot;Number of Polymorphisms&quot;,
         y = &quot;Number of Contigs&quot;) +
    theme(plot.title = element_text(face = &quot;bold&quot;, hjust = .5))
}


#graph read depth distribution
#takes tidy vcf data and optional arguments for title, x limits, ylimits, and fill

graph_depth &lt;- function(data, 
                        my_title = &quot;Depth Distribution&quot;, 
                        xlimits = c(0, 10000), 
                        fill = &quot;#505D83&quot;,
                        ylimits = NULL) {
  data %&gt;% ggplot(aes(x = DP)) + 
  scale_y_log10(limits = ylimits) + 
  geom_histogram(binwidth = 500, fill = fill) + 
  scale_x_continuous(labels = comma, 
                     limits = xlimits) +
  labs(title = my_title, 
       x = &quot;Read Depth&quot;, 
       y = &quot;Number of Polymorphisms&quot;) +
  theme_minimal() +
  theme(plot.title = element_text(hjust = .5, face = &quot;bold&quot;)) 
}


#graph allele balance distribution 
#takes tidy vcf data and optional argument for title, fill, and limits 

graph_allele_balance &lt;- function(data, 
                                 my_title = &quot;Allele Balance Distribution&quot;, 
                                 fill = &quot;#505D83&quot;, 
                                 xlimits = NULL) {
  data %&gt;% ggplot(aes(x = as.numeric(AB))) +
  geom_histogram(binwidth = .05, fill = fill) +
  scale_x_continuous(breaks = c(0, .25, .5, .75, 1),
                     labels = c(&quot;0&quot;, &quot;.25&quot;, &quot;.5&quot;, &quot;.75&quot;, &quot;1&quot;), limits = xlimits) +
  labs(title = my_title, 
       x = &quot;Allele Balance&quot;, 
       y = &quot;Number of Polymorphisms&quot;) +
  theme_minimal() +
  theme(plot.title = element_text(hjust = .5, face = &quot;bold&quot;)) 
}


#graph length distribution of the alternate allele
#takes tidy vcf data and optional argument for title and fill 

graph_length_distribution &lt;- function(data, 
                                      my_title = &quot;Length Distribution&quot;, 
                                      fill = &quot;#505D83&quot;){
  data  %&gt;% 
  filter(str_detect(ALT, &quot;,&quot;) == FALSE) %&gt;%
  mutate(length_alt = nchar(ALT)) %&gt;%
  ggplot() +
  geom_histogram(aes(x = length_alt), binwidth = 2, fill = fill) + 
  scale_y_log10() + 
  theme_minimal() +
  labs(title = my_title,
       subtitle = &quot;Looking at Single-Allelic Sites&quot;,
       x = &quot;Length of Alternate Allele&quot;,
       y = &quot;Number of Polymorphisms&quot;) +
  theme(plot.title = element_text(hjust = .5, face = &quot;bold&quot;),
        plot.subtitle = element_text(hjust = .5)) 
}</code></pre>
<p>Here are example calls to the functions.</p>
<pre class="r"><code>graph_poly_per_contig(spirostomum_668_to_668, &quot;SPIROSTOMUM:\nPolymorphisms per Contig for LKH668 Reads\n Mapped to LKH668 Reference&quot;, fill = &quot;#426DE6&quot;) + ylim(limits = c(0,160)) + xlim(limits = c(0,80))</code></pre>
<p><img src="Basic_VCF_Analysis_files/figure-html/unnamed-chunk-2-1.png" width="672" /></p>
<pre class="r"><code>graph_depth(spirostomum_668_to_668, &quot;SPIROSTOMUM:\nRead Depth Distribution for LKH668 Reads\n Mapped to LKH668 Reference&quot;, fill = &quot;#426DE6&quot;) + xlim(limits = c(0, 6000))</code></pre>
<p><img src="Basic_VCF_Analysis_files/figure-html/unnamed-chunk-2-2.png" width="672" /></p>
<pre class="r"><code>graph_length_distribution(spirostomum_668_to_668, &quot;SPIROSTOMUM:\nPolymorphism Length Distribution for LKH668 Reads\n Mapped to LKH668 Reference&quot;, fill = &quot;#426DE6&quot;) + 
  xlim(0, 90)</code></pre>
<p><img src="Basic_VCF_Analysis_files/figure-html/unnamed-chunk-2-3.png" width="672" /></p>
<pre class="r"><code>graph_allele_balance(spirostomum_668_to_668, &quot;SPIROSTOMUM:\nAllele Balance Distribution for LKH668 Reads\n Mapped to LKH668 Reference&quot;, fill = &quot;#426DE6&quot;) </code></pre>
<p><img src="Basic_VCF_Analysis_files/figure-html/unnamed-chunk-2-4.png" width="672" /></p>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
