<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Vcftools Tutorial</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Analysis of Genetic Diversity in Ciliates</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="Heterozygosity_Tutorial.html">Heterozygosity Tutorial</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Vcftools Tutorial</h1>

</div>


<p>Nucleotide diversity is a measure of genetic variation. It can be used to look at the variation within a population as well as the divergence between populations.</p>
<p>Nucleotide diversity is denoted by π. It is defined as the number of nucleotide differences per site between two randomly selected sequences in a population.</p>
<p>For 2 alleles, the nucleotide diversity is equal to the proportion of different nucleotides times the frequency of each allele in the population. For example, if half the individuals are homozygous G for a variant site while the other half of the individuals are homozygous T, the nucleotide diversity for that site would be ¼ , since 1 * ½ * ½ = ¼ . For more than 2 alleles, nucleotide diversity is equal to the sum of (Freq allele 1) * (freq allele 2) * (nucleotide diversity of the pair) for all pairwise comparisons.</p>
<p>We can calculate nucleotide diversity with <a href="https://vcftools.github.io/man_latest.html">Vcftools</a>. This can be done per variant site, or in windows. We’ll start with the per variant site option.</p>
<p>In the terminal, once you have vcftools installed, you can navigate to the directory where your vcf file is and run a command similar to the following. This produces a tab delimited fite with the suffix ‘.sites.pi’.</p>
<pre class="bash"><code>
#run in the command line
vcftools --vcf input_file_name  --site-pi --out output_file_name
</code></pre>
<p>Once we have the files with the per variant site pi values, we can read them into R.</p>
<pre class="r"><code>library(tidyverse)
library(vcfR)

halt_pi_90 &lt;- read_tsv(&quot;./Data/halteria/halt_pi_90&quot;)
chilo_pi_90 &lt;- read_tsv(&quot;./Data/chilo/chilo_pi_90&quot;)

halt_color = &quot;#964b69&quot;
chilo_color = &quot;#1ea8d6&quot;

chilo_90 &lt;- read.vcfR(&quot;./Data/chilo/all_filtered_chilo_90.vcf&quot;)
halt_90 &lt;- read.vcfR(&quot;./Data/halteria/all_filtered_halt_90.vcf&quot;)</code></pre>
<p>Now we can look at the distribution of per variant site pi values for Halteria and Chilodonella.</p>
<pre class="r"><code>halt_pi_90 %&gt;% 
  ggplot(aes(x = PI)) + 
  geom_histogram(fill = halt_color) + 
  theme_minimal() + 
  labs(x = &quot;Pi&quot;, y = &quot;Frequency (Number of Polymorphisms)&quot;, 
       title = &quot;Halteria Nucleotide Diversity Distribution (MINID90)&quot;) + 
  theme(plot.title = element_text(face = &quot;bold&quot;, hjust = .5))</code></pre>
<p><img src="Nucleotide_diversity_tutorial_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
<pre class="r"><code>chilo_pi_90 %&gt;% 
  ggplot(aes(x = PI)) + 
  geom_histogram(fill = halt_color) + 
  theme_minimal() + 
  labs(x = &quot;Pi&quot;, y = &quot;Frequency (Number of Polymorphisms)&quot;, 
       title = &quot;Chilodonella Nucleotide Diversity Distribution (MINID90)&quot;) + 
  theme(plot.title = element_text(face = &quot;bold&quot;, hjust = .5))</code></pre>
<p><img src="Nucleotide_diversity_tutorial_files/figure-html/unnamed-chunk-3-2.png" width="672" /></p>
<p>When looking at nucleotide diversity, it’s crucial to consider the effect of missing data. Very low and very high π values can results from poorly supported variant sites, and lead to false interpretations of the extent of divergence.</p>
<p>The following function filters for sites that show a variant site in greater than a minimum threshold number of individuals, and that have a genotype called for more than a minimum threshold number of individuals. We can then plot the nucleotide diversity distribution for only the filtered sites.</p>
<pre class="r"><code>#takes vcfR object, minimum number of samples that must have a variant, and minimum number of samples that must have a called genotype 
get_twice_filtered_sites &lt;- function(vcf, min_var = 3, min_geno = 3) {
  ad &lt;- extract.gt(vcf, element = &#39;AD&#39;)
  samples_ad &lt;- ad %&gt;% 
    as_tibble(rownames = &quot;ID&quot;) %&gt;% 
    pivot_longer(cols = 2:(ncol(ad) + 1), names_to = &quot;Sample&quot;) %&gt;% 
    filter(is.na(value) == FALSE) %&gt;% 
    group_by(ID) %&gt;% 
    summarize(num_samples = n()) %&gt;% 
    filter(num_samples &gt;= min_var) %&gt;% 
    pull(ID)
  
  gt &lt;- extract.gt(vcf, element = &#39;GT&#39;)
  samples_gt &lt;- gt %&gt;% 
    as_tibble(rownames = &quot;ID&quot;) %&gt;% 
    pivot_longer(cols = 2:(ncol(ad) + 1), names_to = &quot;Sample&quot;) %&gt;% 
    filter(is.na(value) == FALSE) %&gt;% 
    group_by(ID) %&gt;% 
    summarize(num_samples = n()) %&gt;% 
    filter(num_samples &gt;= min_geno) %&gt;% 
    pull(ID)
  
  intersect(samples_gt, samples_ad)
  
}</code></pre>
<p>Here, we use the function to get the vector of filtered sites for Halteria and Chilodonella. Then we plot the nucleotide diversity distribution for only these sites.</p>
<pre class="r"><code>#get filtered sites
halt_filtered &lt;- get_twice_filtered_sites(halt_90, 
                                          min_var = 3, 
                                          min_geno = 3)

#plot unfiltered distribution for comparison
halt_pi_90 %&gt;% 
  ggplot(aes(x = PI)) + 
  geom_histogram(fill = halt_color) + 
  theme_minimal() + 
  labs(x = &quot;Pi&quot;, y = &quot;Frequency (Number of Polymorphisms)&quot;, 
       title = &quot;Halteria Nucleotide Diversity Distribution (MINID90)&quot;) + 
  theme(plot.title = element_text(face = &quot;bold&quot;, hjust = .5))

#plot filtered distribution
halt_pi_90 %&gt;%
  mutate(ID = paste(CHROM, POS, sep = &quot;_&quot;)) %&gt;% 
  filter(ID %in% halt_filtered) %&gt;%
  ggplot(aes(x = PI)) + 
  geom_histogram(fill = halt_color) + 
  theme_minimal() + 
  labs(x = &quot;Pi&quot;, y = &quot;Frequency (Number of Polymorphisms)&quot;, 
       title = &quot;Filtered Halteria Nucleotide Diversity Distribution (MINID90)&quot;) + 
  theme(plot.title = element_text(face = &quot;bold&quot;, hjust = .5))</code></pre>
<p><img src="Nucleotide_diversity_tutorial_files/figure-html/unnamed-chunk-5-1.png" width="50%" /><img src="Nucleotide_diversity_tutorial_files/figure-html/unnamed-chunk-5-2.png" width="50%" /></p>
<pre class="r"><code>chilo_filtered &lt;- get_twice_filtered_sites(chilo_90, 
                                           min_var = 3, 
                                           min_geno = 3)

chilo_pi_90 %&gt;% 
  ggplot(aes(x = PI)) + 
  geom_histogram(fill = halt_color) + 
  theme_minimal() + 
  labs(x = &quot;Pi&quot;, y = &quot;Frequency (Number of Polymorphisms)&quot;, 
       title = &quot;Chilodonella Nucleotide Diversity Distribution (MINID90)&quot;) + 
  theme(plot.title = element_text(face = &quot;bold&quot;, hjust = .5))

chilo_pi_90 %&gt;% 
  mutate(ID = paste(CHROM, POS, sep = &quot;_&quot;)) %&gt;% 
  filter(ID %in% chilo_filtered) %&gt;%
  ggplot(aes(x = PI)) + 
  geom_histogram(fill = halt_color) + 
  theme_minimal() + 
  labs(x = &quot;Pi&quot;, y = &quot;Frequency (Number of Polymorphisms)&quot;, 
       title = &quot;Chilodonella Nucleotide Diversity Distribution (MINID90)&quot;) + 
  theme(plot.title = element_text(face = &quot;bold&quot;, hjust = .5))</code></pre>
<p><img src="Nucleotide_diversity_tutorial_files/figure-html/unnamed-chunk-6-1.png" width="50%" /><img src="Nucleotide_diversity_tutorial_files/figure-html/unnamed-chunk-6-2.png" width="50%" /></p>
<p>We can also look at nucleotide diversity in windows with Vcftools. This generates a tab delimited file with the suffix ‘.windowed.pi’. After navigating to the directory where your vcf file is, you can run a command like the following. Here, we use a window size of 100 bp.</p>
<pre class="bash"><code>
#run in the command line
vcftools --vcf input_file_name  --window-pi 100 --out output_file_name
</code></pre>
<p>Now we can read these files into R and plot the windowed pi distributions.</p>
<pre class="r"><code>#reading in the tsv files
halt_90_windowed_100 &lt;- read_tsv(&quot;./Data/halteria/halt_window.windowed_90.pi&quot;)
chilo_90_windowed_100 &lt;- read_tsv(&quot;./Data/chilo/chilo_window.windowed_90.pi&quot;)


#plotting the distribution of windowed pi 
halt_90_windowed_100 %&gt;% 
  ggplot(aes(x=PI)) + 
  geom_histogram(bins=60, fill = halt_color) + 
  labs(title = &quot;Halteria:\nDistribution of Windowed Pi (Window Size 100)&quot;, 
       x = &quot;Pi&quot;, 
       y = &quot;Frequency (Number of Windows)&quot;) + 
  theme_minimal() +
  theme(plot.title = element_text(face = &quot;bold&quot;, hjust = .5))</code></pre>
<p><img src="Nucleotide_diversity_tutorial_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
<pre class="r"><code>chilo_90_windowed_100 %&gt;% 
  ggplot(aes(x=PI)) + 
  geom_histogram(bins=60, fill = chilo_color) + 
  labs(title = &quot;Chilodonella:\nDistribution of Windowed Pi (Window Size 100)&quot;, 
       x = &quot;Pi&quot;, 
       y = &quot;Frequency (Number of Windows)&quot;) + 
  theme_minimal() +
  theme(plot.title = element_text(face = &quot;bold&quot;, hjust = .5))</code></pre>
<p><img src="Nucleotide_diversity_tutorial_files/figure-html/unnamed-chunk-8-2.png" width="672" /></p>
<p>We can also learn about the SNP density by looking at the distribution of SNPs per window.</p>
<pre class="r"><code>#Distribution of number of SNPs per window
halt_90_windowed_100 %&gt;%
  ggplot(aes(x=N_VARIANTS)) + 
  geom_histogram(stat = &quot;count&quot;, fill = halt_color) +
  scale_x_continuous(limits = c(0,15), breaks = seq(0,25, by =1))  +
  labs(title = &quot;Halteria:\nDistribution of Number of SNPs Per Window\n (Window Size 100)&quot;, 
       x = &quot;Number of SNPs&quot;, 
       y = &quot;Frequency (Number of Bins)&quot;) + 
  theme_minimal() +
  theme(plot.title = element_text(face = &quot;bold&quot;, hjust = .5))</code></pre>
<p><img src="Nucleotide_diversity_tutorial_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
<pre class="r"><code>chilo_90_windowed_100 %&gt;%
  ggplot(aes(x=N_VARIANTS)) + 
  geom_histogram(stat = &quot;count&quot;, fill = chilo_color) +
  scale_x_continuous(limits = c(0,15), breaks = seq(0,25, by =1))  +
  labs(title = &quot;Chilodonella:\nDistribution of Number of SNPs Per Window\n (Window Size 100)&quot;, 
       x = &quot;Number of SNPs&quot;, 
       y = &quot;Frequency (Number of Bins)&quot;) + 
  theme_minimal() +
  theme(plot.title = element_text(face = &quot;bold&quot;, hjust = .5))</code></pre>
<p><img src="Nucleotide_diversity_tutorial_files/figure-html/unnamed-chunk-9-2.png" width="672" /></p>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
